File: 919-complete-binary-tree-inserter/919-complete-binary-tree-inserter.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class CBTInserter:
    
    def __init__(self, root: Optional[TreeNode]):
        self.root = root

    def insert(self, val: int) -> int:
        queue = deque()
        queue.append(self.root)
        
        node = TreeNode(val)
        
        while queue:
            currNode = queue.popleft()
            
            if currNode.left == None:
                currNode.left = node
                return currNode.val
            else:
                queue.append(currNode.left)
                
            if currNode.right == None:
                currNode.right = node
                return currNode.val
            else:
                queue.append(currNode.right)
            
            
    def get_root(self) -> Optional[TreeNode]:
        return self.root


# Your CBTInserter object will be instantiated and called as such:
# obj = CBTInserter(root)
# param_1 = obj.insert(val)
# param_2 = obj.get_root()

File: 919-complete-binary-tree-inserter/919-complete-binary-tree-inserter.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class CBTInserter {
    TreeNode* root_;
public:
    CBTInserter(TreeNode* root) {
        root_ = root;
    }
    
    int insert(int val) {
        TreeNode* node = new TreeNode(val);
        
        queue<TreeNode*> queue;
        
        queue.push(root_);
        
        while(!queue.empty()) {
            auto currNode = queue.front();
            queue.pop();
            
            if(currNode -> left == nullptr){
                currNode -> left = node;
                return currNode -> val;
            }
            else
                queue.push(currNode -> left);
            
            if(currNode -> right == nullptr){
                currNode -> right = node;
                return currNode -> val;
            }
            else
                queue.push(currNode -> right);
                                    
        }
        return -1;
            
    }
    
    TreeNode* get_root() {
        return root_;
    }
};

/**
 * Your CBTInserter object will be instantiated and called as such:
 * CBTInserter* obj = new CBTInserter(root);
 * int param_1 = obj->insert(val);
 * TreeNode* param_2 = obj->get_root();
 */